### DISCUSSION TITLE
VE.Direkt Verbindung

### DISCUSSION THREAD

**Discussion by lke15** [hoylabs/OpenDTU-OnBattery#47] [Q&A] [answered]
Hallo,
wie kommuniziert der ESP mit dem MPPT Solar Charger? 
Per Funk/Bluetooth oder per Kabel wie beim "Victron MPPT ve.direct to MQTT" Projekt.

**Comment by flofresser** [comment, marked as answer]
Du Musst TX, RX und GND vom Victron mit dem ESP32 Verkabeln.
TX und RX müssen gekreuzt sein.
Beim Programm wurden folgende GPIO‘s verwendet:

-DVICTRON_PIN_TX=21
-DVICTRON_PIN_RX=22

**Reply by dbohne** [reply]
hi zusammen, ich verstehe irgendwie nicht wie das gemeint ist "TX und RX müssen gekreuzt sein." Kann ich denn nicht einfach per Kabel die pins verbinden? Brauche ich denn noch ein Modul dazwischen ?Vielen Dank

**Reply by madmartin** [reply]
Damit ist gemeint: auf der einen Seite das TX wird mit dem RX der anderen Seite verbunden und umgekehrt. Ist wie mit Mund und Ohr - was auf der einen Seite gesprochen/gesendet wird, muß auf der anderen Seite gehört/empfangen werden.

Die Frage nach dem Modul ist interessant, da ich das selbst erst in kürze aufbauen werde kann ich das nicht sicher beantworten. Die Doku zum Hardwareaufbau ist ja gerade in Arbeit.

@helgeerbe ???

**Reply by helgeerbe** [reply]
Schaut mal hier #138

Victron ist 5v ESP 3.3V. Dauerhaft würde ich einen Level Shifter dazwischen setzen. Zum Testen sollte der ESP es vertragen

**Reply by dbohne** [reply]
Vielen Dank. Dann weiß ich jetzt was zu tun ist 🙏

**Reply by dbohne** [reply]
ich bekomme es irgendwie nicht zum laufen, verdammt. Ich habe ein ESP32 wroom.32. welche pins muss ich dafür nurtzen? Sorry für die blöde Frage. Brauche ich da auch den Level Shifter? Vielen Dank und Kompliment

**Reply by helgeerbe** [reply]
Ich verwende die Pins 21, und 22 bei meinem D1 Mini. Aber du kannst die Pins auch über das Pin Mapping zuweisen. Bitte mache für dein Problem ein eigenes Thema auf. Ansonsten fällt es mir schwer den ganzen Diskussionen, zu folgen.

**Comment by Adminius** [comment]
warum eigentlich 21 und 22? diese sind standardmäßig für i2c definiert. 
16 und 17 wären so standard UART Pins (HW Serial 2).
ja, man kann die Pins frei zuweisen. Ich wundere mich halt. 

und wenn wir schon dabei sind: da man beim Victron eigentlich nur TX braucht da wir eigentlich nur lesen, könnte man RX pin für die Spezielle Victron Funktionen gleich verwenden/auswerten (z.B. Virtueller Last Ausgang). Dann müsste man als TX pin irgend einen "falschen" pin vergeben, und ESP32 TX Pin als IO Pin nutzen...

**Comment by helgeerbe** [comment]
Ich hatte erst einen esp826 und mir eine ve.direct <-> mqtt Bridge gebastelt. Als dann OpenDTU kam, habe ich einfach einen ESP32 D1 Mini in die Fassung gesteckt. Dadurch stehen mir nicht alle PINS zur Verfügung. OpenDTU hatte Pin 16 und 17 schon in Verwendung. 21 und 22 sind Zufall. Allerdings stehen sie direct unter den TX/RX Pins.  (Wenn du dir mal das Pinout vom D1 Mini esp32 anschaust).

RX habe ich verbunden, falls ich mal Langeweile habe. Dann würde ich statt des Ascii Protokoll auch das HEX Protokoll implementieren (dazu bracht man RX). Allerdings war meine Langeweile noch nie so groß, dass ich es in Angriff genommen hätte.

**Reply by Adminius** [reply]
also historische Gründe... ich nutze selbst esp32 D1 mini, ich habe aber gleich darauf geachtet, welche pins ich für welche Sachen freihalte :) 
na gut, da man ja jetzt "einfacher" einstellen kann und nicht jedes mal neu kompilieren muss, hat sich das eh erledigt :)
